### ISSUE TITLE
Standardize on invokation of callables?

### ISSUE THREAD

**Issue by mindplay-dk** [container-interop/container-interop#43]
Can we standardize on a facet/interface that defines invokation of callables?

For example, in a couple of projects, I define a very similar interface (and matching exception type) along the lines of these:

```php
/**
 * This interface defines the call signature of a component capable of
 * invoking a callable, given a set of named parameters.
 *
 * By implementing this interface, you can create a custom invoker to
 * provide integration with e.g. a dependency injection container.
 */
interface InvokerInterface
{
    /**
     * @param callable $func   the Closure to be invoked
     * @param array    $params map where parameter name => value
     *
     * @return mixed return value from the invoked function
     *
     * @throws InvocationException on missing parameter
     */
    public function invoke(callable $func, array $params);
}

/**
 * This Exception is thrown if there is a problem invoking a callable.
 *
 * @see Invoker::invoke()
 */
class InvocationException extends FunctionException
{
}
```

The `invoke()` function takes a set of parameter name/value overrides, and must resolve the rest by whatever means are implemented by the underlying container. The invoker should fill optional arguments with default values, when available, and should throw an `InvocationException`.

The purpose of this interface is to provide a means of switching the behavior from a default, built-in argument filler to a real DI container. Currently, each of these projects need a similar integration adapters for each individual DI container. A standardized interface would allow these projects to interop with any container that supports this interface.

Because it's an interface, implementors can choose to implement this as a facet of the container itself, or as a dedicated, separate component.

Thoughts?

**Comment by mnapoli** [comment]
:+1: I'm 100% behind this. PHP-DI has a similar feature ([scroll down to `call()`](http://php-di.org/doc/container.html)) and it got so useful for other projects of mine that I ended up extracting that in its own package: [PHP-DI/Invoker](https://github.com/PHP-DI/Invoker).

There is the question however of whether there should be the `callable` type-hint.  In my implementation, the invoker has 2 features:

- resolve the callable
- invoke the callable

The first one is interesting because it allows to [store some callables in a DI container](https://github.com/PHP-DI/Invoker#resolving-callables-from-a-container). That means you can invoke a container entry: `$invoker->invoke(['my.super.service', 'method'], $params)`. To illustrate that this is exactly what happens in frameworks: in the router you can either define controllers as callables (closures, …), or controller as services. The invoker becomes the dispatcher!

I've run into issues however where sometimes you only want to resolve the callable and not invoke it, so I started working on separating that in its own class (https://github.com/PHP-DI/Invoker/pull/2). Maybe that's a sign that it's a mix of SRP violation and it's better to have those 2 responsibilities separated…

However encapsulating all that behavior behind a single interface allows to extend very easily libraries or framework to support e.g. [extended callables](https://gist.github.com/mnapoli/f8afef0f556010c144e7) which is super useful.

**Comment by njasm** [comment]
I'm not yet buying the idea. For what i see (and i'm sorry if i missunderstanding it) .

Can you give a more real world example why to standard it with an interface? 
Thanks

**Comment by mindplay-dk** [comment]
@njasm 

> Can you give a more real world example why to standard it with an interface? 

Yeah, [this router](https://github.com/mindplay-dk/walkway) - it uses a hiearchy of callback functions for hierarchical dispatch... so, for example, given URLs like `post/123/edit` and `post/123/delete`, the `post/123` callback might resolve `Post` instance with the primary key `123` for both the `edit` and `delete` sub-paths. It has a built-in simple invoker, which is able to inject only things that were explicitly injected by previous paths, or at initialization - with this interface, it would be possible to inject e.g. `PostRepository` from a compatible DI container.

Another example is [this simple view-engine](https://github.com/mindplay-dk/kisstpl), which can inject certain things into template closures - in this case, coupling to a DI container via this interface would permit injection of e.g. various URL or view-helpers etc. registered in the container.

Another example could be various routers or micro-frameworks, where URL patterns get mapped to closures - if these could plug and play with any container, they'd be more directly useful without having to build adapters and "frameworky" stuff around them first...

**Comment by mindplay-dk** [comment]
@mnapoli I noticed while I was using PHP-DI that you invoke method also is capable of resolving component names in the container. I think PHP-DI is the first and only container I've seen that does that, so there is probably no real need to standardize on that? (Which wouldn't prevent you from supporting this interface, you'd just have to do it with an adapter instead of implementing it directly.)

**Comment by Ocramius** [comment]
The interface above is really weird: are you trying to standardize a functor? In OO-world, that pretty much is a factory.

**Comment by mnapoli** [comment]
@Ocramius I honestly don't think I understood what you wrote :p It's not a factory in the sense that it's not meant to create something, it's rather meant to abstract how something should be called (e.g. what parameters are passed to the callable).

In my case, just like @mindplay-dk (I love that router btw) I use some `InvokerInterface` to invoke the controllers in a micro-framework (https://github.com/stratifyphp/framework/blob/master/src/Middleware/ControllerInvoker.php#L63) as well as invoke the commands (aka CLI controller) in a CLI micro-framework (http://mnapoli.fr/silly/docs/dependency-injection.html).

More explicitly, it can allow that:

```php
$app->post('/register/{name}', function ($name, Mailer $mailer) {
    $mailer->sendMail($name, 'Welcome!');

    return 'You have received a new email';
});
```

@mindplay-dk "resolving callables from the container" may not be a common feature in DI containers, but it's a common pattern in frameworks, e.g. in Silex, Slim, Zend Expressive, Silly, … Each implements its own "callable/controller resolver". Anyway maybe this is getting off topic.

**Comment by Ocramius** [comment]
@mnapoli forget it, I had a brain fart and realized what this is about much later. I'm just not used to call-time parameter resolution (I never do it)

**Comment by mindplay-dk** [comment]
@mnapoli good code example - and looks like [Auryn has this feature](https://github.com/rdlowrey/Auryn/blob/master/lib/Injector.php#L611) and [Laravel does too](https://github.com/illuminate/container/blob/master/Container.php#L495), though with a slightly different interface, but could interop via an adapter. I think this is a pretty common feature in DI containers?

**Comment by mnapoli** [comment]
Interesting, both Auryn and Laravel accept *non-callables* too:

- `Class@method` for Laravel
- `Class::method` for Auryn

Both will resolve `Class` from the container (it seems).

**Comment by mindplay-dk** [comment]
> both Auryn and Laravel accept non-callables too

As I understand, e.g. `$container->invoke([$container->get(Foo::class), 'bar'])` is functionally equivalent to `$container->invoke('Foo::bar')`, right?

I'd prefer the first - it's more verbose, but doesn't require you to invent a hybrid extension of `callable`, it's more explicit about what it does, and it has static coupling at least to the class-name.

Beyond convenience, yes, `'Foo::bar'` can be stored, whereas `[$container->get(Foo::class), 'bar']` would evaluate and initialize `Foo` immediately, so that's one argument in favor - however, if that's a pressing need, I think I'd prefer to standardize on something like a [boxed value](https://github.com/mindplay-dk/unbox/blob/master/src/BoxedValueInterface.php) interface, which would permit any container to provide a lazy reference to a service for deferred initialization.

**Comment by mnapoli** [comment]
Here is the difference for example with routing:

```php
$app->get('/blog', 'Acme\BlogController::index');
$app->get('/blog', [$app->getContainer()->get('Acme\BlogController'), 'index']);
```

Not very appealing for users, even when ignoring the lazy-loading problems.

And if routing is defined in XML or YAML (or anything not-PHP), it's not possible.

**Comment by Ocramius** [comment]
It would be wise to NOT rely on a magic sequence of characters ( `::` ) for
this sort of logic.
You may want to loop @danack into this, as he is currently trying to fix
the horrible thing that PHP callables are right now:
https://wiki.php.net/rfc/consistent_callables

Marco Pivetta

http://twitter.com/Ocramius

http://ocramius.github.com/

**Comment by mnapoli** [comment]
What's magic is the resolution of 'Acme\BlogController' (or could be `my_blog_controller`) through the container, not a sequence of characters. At least that's why I suggest. Symfony has the magic `:` (instead of `::`) to do controller-as-service. Laravel uses `@` as a separator.

**Comment by Ocramius** [comment]
Both valid for class names in my world, eh :-P

**Comment by mindplay-dk** [comment]
> Not very appealing for users, even when ignoring the lazy-loading problems.

Note I use `ref()` rather than `get()` in these situations - it returns a lazy "boxed reference" rather than initializing the service.

> And if routing is defined in XML or YAML (or anything not-PHP), it's not possible.

Of course it is - you could have the *loader* support this via syntax or directives etc, rather than breaking `callable`. Alternatively, your Router API could accept an overloaded `callable` or provide additional methods etc... there are plenty of ways to solve this problem without overloading `callable`.

Sorry, but I don't buy that overloading `callable` is necessary, and I don't believe that overloading internal pseudo-types is as good idea.

**Comment by mnapoli** [comment]
> Of course it is - you could have the loader support this via syntax or directives etc

Yep, but we're back to "custom representation of a callable" (or at least a controller).

> I don't believe that overloading internal pseudo-types is as good idea.

Fair enough I understand. My point is that this is already what most framework do but in a not so obvious way when doing stuff like `Controller:method` (note the use of `:` and not `::`) or `Controller@method`, etc. Each framework has its format of representing an "improved" callable (because if we are honest that's what's represented here). I'd rather see a unique format across the board.

Then there is the question of whether or not it's the responsibility of the `Invoker` to resolve those "custom" callables (open question, but separate from the other discussion).

But let's leave it at that since the idea doesn't have a lot of support today.

**Comment by mindplay-dk** [comment]
@mnapoli I think the method name alone implies there's something wrong - assuming the method name is e.g. `invoke()` and not `invokeOrResolveAndInvoke()`... it definitely seems like two separate responsibilities - I'd definitely argue that resolving has nothing directly to do with invocation. (though come to think of it, that also raises the question of whether invocation is a concern that directly has anything to do with container interoperability in the first place... hmm...)

**Comment by glen-84** [comment]
What is required to move this proposal forward?

FWIW, I don't think that it should resolve services as well, that can be a separate method, or you can just resolve the service first and pass that in.

I agree with the signature `(callable $callable, array $params)`.

Would `call` work better than `invoke`? (it's short, and possibly used more than `invoke` in the PHP world)

Also, shouldn't this be in the container interface? If it's a separate interface then you can't be sure that an implementation will support it (like `has` and `get`).

I'm not sure what the processes are, but if you want me to set up some polls for the above questions, let me know.

**Comment by mnapoli** [comment]
@glen-84 see https://github.com/container-interop/container-interop/issues/44#issuecomment-185742307

> Also, shouldn't this be in the container interface?

No it shouldn't (https://en.wikipedia.org/wiki/Interface_segregation_principle), and there's no need for get() and has().

**Comment by mindplay-dk** [comment]
> No it shouldn't 

I agree, this should be a separate interface.

The case where something depends on both `ContainerInterface` and `FactoryInterface` is correctly handled with a signature like `__construct(ContainerInterface $container, FactoryInterface $factory)`, even in cases where those two arguments are the same container instance - to the consumer, these are two different responsibilities, and they may or may not be provided by the same object, that's not really for the consumer to dictate, but instead something the provider of those two dependencies should decide.

Having two interfaces will also simplify testing of units that depend on both facets, as you can mock one, or the other, or both - again, consumer's choice.

Also, as a container vendor, I can opt to support one, or the other, or both facets - implement them in the same class, or in separate classes, etc... two interfaces afford by far the most flexibility.

**Comment by glen-84** [comment]
I guess I just saw them as a set of closely-related methods. Anyway, point taken.

My use case involves the use of both a container-aware factory *and* an invoker, but it seems as though it will be impossible for me to do this in an interoperable way. :disappointed:

**Comment by mindplay-dk** [comment]
> it seems as though it will be impossible for me to do this in an interoperable way

Not at all:

```php
class YourClass {
    /** @var ContainerInterface */
    private $container;

    /** @var InvokerInterface */
    private $invoker;

    public function __construct(ContainerInterface $container, InvokerInterface $invoker) {
        $this->container = $container;
        $this->invoker = $invoker;
    }

    public function stuff() {
        $a = $this->container->get(...);
        $b = $this->invoker->invoke(...);
    }

    // ...
}

class YourContainer implements ContainerInterface, InvokerInterface {
    // ...
}

// your app:

$container = new YourContainer();

$your_object = new YourClass($container, $container); // see?
```

Two responsibilities, one object.

**Comment by glen-84** [comment]
@mindplay-dk My apologies, I wasn't very clear. What I meant was that I can't get this to work, because the interfaces do not exist yet, and C-I is not interested unless the big-name frameworks and libraries start banging at their door, asking for them.

**Comment by mindplay-dk** [comment]
Well, if they existed already, we wouldn't be having this conversation ;-)

**Comment by mnapoli** [comment]
@glen-84 A good thing you could do is try to talk about this interface to your framework maintainers and try to get them interested in the idea?

I tried to pitch the "invoker" interface to Slim when the v3 was in the works but I didn't see too much enthusiasm (using ContainerInterface was already a great step).

**Comment by mindplay-dk** [comment]
It's not only up to maintainers - third parties can (easily) create interop
layers/adapters, even when maintainers don't care enough to do it
themselves.

Adoption doesn't have to come from the framework maintainers exclusively.
It has the same value to those who want it, regardless of who creates it.

**Comment by weierophinney** [comment]
We did something similar to this in ZF2's `Zend\Stdlib\CallbackHandler`. As such, I want to throw one piece of information out there: wrapping callables like this can have significant performance impact. We were surprised when we saw a 25-40% improvement simply by removing the `CallbackHandler` wrappers (in favor of simple callables) in zend-eventmanager. If we needed lazy-loading, the PHP engine was far more efficient at invoking closures such as the following:

```php
function ($event) use ($container) {
    $callback = $container->get(ServiceName::class);
    $callback($event);
}
```

than it was with using the OO wrapper.

**Comment by mindplay-dk** [comment]
@weierophinney even with apparently "huge" gains of 25-40%, we're probably talking about micro-optimizations either way? with PHP, it used to be "every bit counts", but since 5.6 and 7.0, not so much anymore. In some combined benchmarks with a full stack (dispatching real controllers, making a DB round-trip, etc.) the DI container overhead was something like 0.5% of total request overhead, so net savings of even 50% equates to like 0.25% of a typical request cycle, which really isn't worth worrying about, especially now with php 7 ;-)

**Comment by glen-84** [comment]
@mindplay-dk Apparently it was 10+% for PHP7, but I'm personally not concerned with that in cases where you may only be invoking a single method like this during a request.

**Comment by mnapoli** [comment]
@weierophinney you are talking about wrapping callables in objects that will lazily resolve the target to invoke right?

What was suggested originally is a "service" (i.e. stateless object) that would invoke callables with the given parameters, and resolve missing parameters (in most cases from a container). Usage example:

```php
$invoker->invoke(function ($name, Mailer $mailer) {
    $mailer->sendMail($name, 'Welcome!');
}, [
    'name' => 'Gob',
]);
```

Regarding performances, in your use case (event dispatcher) there would be no creation of plenty of callable-wrapper objects, so there would be no overhead here. However there might be overhead when invoking listeners because it goes through the invoker (obviously).

But I'm not even sure why you would need an invoker object as described here since (I guess) you don't have dynamic parameters in event listeners? Maybe the use case you mentioned is only about *lazy loading* the callables?

If that's the case then the `callable` type-hint of the `invoke()` method would prevent to do that (because service names for example would be strings and not callables). However that's still a very interesting use case (event listeners, controllers, CLI commands, middlewares, etc.), I mentioned allowing non-callables as the first parameter of `invoke()` in that comment: https://github.com/container-interop/container-interop/issues/43#issuecomment-149006563 However it was pointed it that resolving a callable might be a separate responsibility from invoking it, and now I tend to agree with that (e.g. https://github.com/PHP-DI/Invoker/pull/5).

I don't know however if the PHP ecosystem would benefit from a "callable resolver interface"… The good thing is that it may allow dispatchers (routers), micro-frameworks, CLI applications (https://github.com/symfony/symfony/pull/13946), event dispatchers, command bus, etc. to be decoupled from containers. But it's often possible to write adapters though (but not always).
